### DISCUSSION TITLE
What is the difference between the MyFile.dll and MyFile.CA.dll generated when building a Custom Action from the WiX template?

### DISCUSSION THREAD

**Discussion by MattWalker414** [wixtoolset/issues#7762] [Questions] [answered]
I've had some issues pointed out to me that Windows Defender is flagging some binaries related to Custom Actions during our installation, authored using InstallShield, but some MSI dll Custom Actions are written using the WiX template (C#).

I've decided to try to sign all of my Custom Action executables, but when I build the MSI dll CA, I get two .dlls in the output folder...

MyFileCheck.dll
MyFileCheck.CA.dll (smaller in size)

When I compile the dll in Visual Studio, only the MyFile.dll gets signed.  Since in InstallShield, the Custom Action is configured by pointing to MyFile.CA.dll, I tried to sign that binary as well, but that presents errors...

EXEC : SignTool error : SignedCode::Sign returned error: 0x800700C1
1>  
1>  	Either the file being signed or one of the DLL specified by /j switch is not a valid Win32 application.
1>  
1>EXEC : SignTool error : An error occurred while attempting to sign: C:\Install Source\MyApp\Custom Actions\MyFileCheck\MyFileCheck\bin\Release\MyFileCheck.CA.dll

Again, with using InstallShield, which dll is used from/included in the compiled installation pacakge?  

The command im using to sign is....
      signtool sign /f my.pfx /p mypswd $(TargetDir)$(TargetFileName)

**Comment by pas059** [comment]
Hi,
as i know the xxx.dll file contains CLR code, and the xxx.CA.dll contains binary.
regards

**Reply by MattWalker414** [reply]
Thanks pas059!  I think I tried setting the InstallShield custom action to use the xxx.dll and not the xxx.CA.dll and that errored.  I believe at runtime.  I'm still not sure I understand and if the signing of the xxx.dll will suffice for Defender if the install actually uses the xxx.CA.dll.

**Reply by chrpai** [reply]
Here's a blog article I wrote 15 years ago when DTF first came out.  It's very comprehensive.  The only part to ignore is running a postbuild step to call sfxca.  This was added to the .wixproj soon after.

http://blog.deploymentengineering.com/2008/05/deployment-tools-foundation-dtf-custom.html

Build an InstallShield Basic MSI with an InstallScript custom action.  Export the  ISSetup.dll from the Binary table and look at it in Depends.  You'll see it has a bunch of exported functions caled F1 F2 F3 and so on.  If you look at the CustomAction table you'll see how they map.   Now do the same with a WiX DTF CA.   Compare that and what my blog post explains and it should all make sense.

**Comment by chrpai** [comment]
For signing, $(TargetFileName) will get you the .dll.  You'll want to hard the DLL.CA.DLL  or use another MSBuild property.  You probably also want to specify a timestamp server and the digest alogs for signing and timestamp signing.

**Reply by MattWalker414** [reply]
Yes, TargetFileName gets the dll.

I did hard code the CS.dll in post events...

signtool sign /f my.pfx /p mypswd (TargetDir)\MyFileCheck.CA.dll

..., but that is what generates the errors.  Is that maybe because of the missing/other attributes you specified?

**Reply by chrpai** [reply]
I don't see $ infront of $(TargetDir).  I can't tell you what is causing your error without knowing what the error message is.

**Reply by MattWalker414** [reply]
Yeah, the $ was stripped when I pasted the stuff in.  The syntax of the command is correct I believe and the error is in the original post.

**Reply by chrpai** [reply]
A quick google search for signtool and that error code suggests  a Windows PE file with a bad header (doubt it)  or using the wrong bitness of signtool for the file.   Make sure you are calling an x86 signtool if your signing an x86 dll.

**Reply by MattWalker414** [reply]
I think I checked into bitness, but I'll double back.  I'm using the same signtool.exe for the regular .dll and that signs successfully.

I guess the first real question is, is the .CA.dll a valid library that is signable?

**Reply by jsrustad** [reply]
*.CA.dll is a real library that is signable.  What version of WiX are you using?

**Reply by MattWalker414** [reply]
3.something I believe.

**Reply by barnson** [reply]
"3.something" covers a period of 14 years...

**Reply by jsrustad** [reply]
I suspect you're seeing the problem I described at https://github.com/orgs/wixtoolset/discussions/7762#discussioncomment-7149063

**Reply by MattWalker414** [reply]
3.11, but I see the certificate of the .dll expired 4/7/22.  I wonder if that is the problem.

**Reply by jsrustad** [reply]
3.11.1 does not have the problem.
3.11.2 does have the problem.
The expiration date on the certificate is not a problem.  The signature has a timestamp indicating when it was signed.  As long as that timestamp is within the valid range on the certificate, the signature is valid.
Try the fix I recommended in my proposed answer.

**Reply by MattWalker414** [reply]
That appeared to work.  I removed the cert from the x86 template .dll and my .CA.dll was then signed successfully!  Thanks for that.
I would imagine if I have x64 .dlls the x64 template is used.  Can I remove the cert using the same tool since I compiled that for x86 or does that have to be compiled to x64 as well?  What about actions compiled as Any CPU, which template is used in that case?

**Reply by jsrustad** [reply]
I do not know the answers to those questions as I have never dealt with x64 installs.

**Reply by MattWalker414** [reply]
I hope I can get some answers as I don't want to do something that might look successful but fails in the field.

**Reply by MattWalker414** [reply]
Does anyone know... since it appears that I've run into a bad custom action .dll template in WiX (bad signature) where Delcert needs to be used - https://github.com/MadhukarMoogala/delcert...  if I am signing an x64 custom action, do I need to recompile the delcert tool in x64 to remove the bad signature from the x64 template .dll?  Also, what about Any CPU code?  What template file is used to create that type of .dll?

**Reply by chrpai** [reply]
While I haven't read every word of this thread, I suspect your over thinking this.   On your machine take a look at these files
C:\Program Files (x86)\WiX Toolset v3.11\SDK\x86\sfxca.dll
C:\Program Files (x86)\WiX Toolset v3.11\SDK\x64\sfxca.dll

You will see they are signed.  You will see the signature times are a little bit different. Note their file size.  When you build your custom action project, you should see the signature still there but invalid because the file has grown with your custom action code.

The rub is these files should have never been signed.  If you use delcert on these two files your problem should go away.  The output files will not be signed and then signtool will work.

Or just call delcert before calling signtool.

Or just use wix 3.14  or wix 4.0.

Do you need to recompile delcert? I don't know... I've never used it.   Why not just compile an x64 and x86 custom action and give it a try.  See if it throws an error and look at the file properties to see if it worked.

**Reply by MattWalker414** [reply]
I guess that is basically my question, do I need to recompile delcert as 64 bit to remove the signature from the 64 bit template dll?  I can just give that a try with the x86 delcert.

Initially, I had inadvertently tried to recompile delcert as 64 bit and received a host of errors.

**Reply by chrpai** [reply]
Maybe you should ask @MadhukarMoogala.

**Reply by jsrustad** [reply]
I don't think you need to recompile delcert as 64-bit.  Try it and see.

Since I gave you the answer, please accept my answer.

**Reply by MattWalker414** [reply]
Yes, I was definitely over thinking it!

**Comment by jsrustad** [comment, marked as answer]
Which version of WiX are you using?

WiX 3.11.2 had a problem where the sfxca.dll template was signed. When your MyFile.dll was added to this template to make MyFile.CA.dll, the signature was invalidated. Microsoft's signtool won't sign a file that has an invalid signature. Signtool also won't remove an invalid signature. I only know this because I was bitten by this myself.
To work around this problem, use delcert to remove the signature from the template file.
The bug report for WiX - https://github.com/wixtoolset/issues/issues/6089
Delcert is available at - https://github.com/MadhukarMoogala/delcert

**Reply by MattWalker414** [reply]
It all worked out just compiling delcert to x86 and running against both the x86 and x64 WiX custom action template .dlls.

Thanks so much for this info!

**Reply by MattWalker414** [reply]
So, I'm finally getting back to working to ensure that WiX custom action dlls included in and used by my installs are signed properly.  I thought all of the steps were here, but someone had mentioned that these DLLs are not signed properly with Visual Studio post build steps because the actual .dll is already encapsulated in the CA dll at sign time and only the CA.dll is signed.  There were a series of post build steps to get around that, but I can for the life of me find the info.  I thought those steps were here, but they must be in another thread.

The steps were (and I may have the main and CA.dlls mixed up) to add build steps to delete the newly created .exe, recreate the main .dll, sign it, then run the Wix widget to encapsulate that .dll in the CA.dll or something to that effect.

**Reply by jsrustad** [reply]
You build MyFile.dll, which is then added to the sfxca.dll template to come up with MyFile.CA.dll.  You then Authenticode sign MyFile.CA.dll.  Why do you think you also need to sign MyFile.dll?

When you run the installer, it extracts MyFile.CA.dll from a binary stream in the installer into a temporary file.  The installer then calls the entry point function on the temporary file, which runs your custom action.

**Reply by MattWalker414** [reply]
At some point I thought someone said that compiling through Visual Studio and signing using a post build script caused an issue.  The encapsulated .dll would not be signed post compile as its already encapsulated in the CA.dll, which does get signed.  I guess that might trip defender.  I thought all of that information was here, as someone spelled out all of the post build steps to get it right, but it must be in a separate thread.  I can't for the life of me find it.

**Reply by jsrustad** [reply]
You are correct that the embedded DLL (MyFile.dll) doesn't get signed.  I don't think that will be a problem because the outer DLL (MyFile.CA.dll) is signed.  The only way to tell for sure is to do some testing.  I can say that I have not seen a problem with the embedded DLL not being signed.

**Reply by MattWalker414** [reply]
I believe whoever provided the steps to recompile to have both .dlls signed did so because they say issues with threat detection.  I really wish I could find that posting.  I was certain that info was here, but I'll keep digging.

**Reply by MattWalker414** [reply]
Found it!  https://community.flexera.com/t5/InstallShield-Forum/Custom-Action-Utilities-Flagged-By-Windows-Defender/m-p/295625

**Reply by jsrustad** [reply]
Interesting.  I will squirrel this away in case I need it.

**Comment by chrpai** [comment]
Thanks, I wasn't aware of that bug.  I had moved onto the dev build 3.14 years ago due to other important fixes to me.  I really wish a final 3.x would have shipped.  It's nice to have an old stable with all known bug fixes and a v.next to choose from.

https://wixtoolset.org/docs/v3/releases/v3-14-0-6526/
